godot-core/src/builtin/callable.rs

```diff
@@ -0,0 +1,178 @@
+///
+/// [godot-cpp#802]: https://github.com/godotengine/godot-cpp/issues/802
+#[repr(C, align(8))]
+pub struct Callable {
+    opaque: sys::types::OpaqueCallable,
+}
+
+impl Callable {
+    fn from_opaque(opaque: sys::types::OpaqueCallable) -> Self {
+        Self { opaque }
+    }
+
+    /// Create a callable for the method named `method` in the specified `object`.
+    ///
+    /// _Godot equivalent: `Callable(Object object, StringName method)`_
+    pub fn from_object_method<T, S>(object: Gd<T>, method: S) -> Self
+    where
+        T: GodotClass, // + Inherits<Object>,
+        S: Into<StringName>,
+    {
+        // upcast not needed
+        let method = method.into();
+        unsafe {
+            Self::from_sys_init_default(|self_ptr| {
+                let ctor = sys::builtin_fn!(callable_from_object_method);
+                let args = [object.sys_const(), method.sys_const()];
+                ctor(self_ptr, args.as_ptr());
+            })
+        }
+    }
+
+    /// Calls the method represented by this callable. Arguments passed should match the method's
+    /// signature. If called with more arguments than expected by the method, the extra arguments will be
+    /// ignored. If called with fewer arguments than expected it will crash. If called with arguments of the
+    /// wrong type then an error will be printed and the call fails.
+    ///
+    /// _Godot equivalent: `callv`_
+    pub fn callv(&self, arguments: VariantArray) -> Variant {
+        self.as_inner().callv(arguments)
+    }
+
+    /// Returns the name of the method represented by this callable. If the callable is a lambda function,
+    /// returns the function's name.
+    ///
+    /// _Godot equivalent: `get_method`_
+    pub fn method_name(&self) -> Option<StringName> {
+        let method_name = self.as_inner().get_method();
+        if method_name.is_empty() {
+            None
+        } else {
+            Some(method_name)
+        }
+    }
+
+    /// Returns the object on which this callable is called.
+    ///
+    /// _Godot equivalent: `get_object`_
+    pub fn object(&self) -> Option<Gd<Object>> {
+        // Increment refcount because we're getting a reference, and `InnerCallable::get_object` doesn't
+        // increment the refcount.
+        self.as_inner().get_object().map(|object| {
+            <<Object as GodotClass>::Mem as crate::obj::mem::Memory>::maybe_inc_ref(&object);
+            object
+        })
+    }
+
+    /// Returns the ID of this callable's object, see also [`Gd::instance_id`].
+    ///
+    /// _Godot equivalent: `get_object_id`_
+    pub fn object_id(&self) -> Option<InstanceId> {
+        let id = self.as_inner().get_object_id();
+        InstanceId::try_from_i64(id)
+    }
+
+    /// Returns the 32-bit hash value of this callable's object.
+    ///
+    /// _Godot equivalent: `hash`_
+    pub fn hash(&self) -> u32 {
+        self.as_inner().hash().try_into().unwrap()
+    }
```

Comment:
Is this guaranteed not to panic? The values returned by Godot are unsigned?

Reply:
according to [`Array::hash`](https://github.com/godot-rust/gdext/blob/885bb91926c76867093d85208b33aa80c87183b6/godot-core/src/builtin/array.rs#L95) at least yeah. i think i checked this myself at some point too.

Reply:
also if this ever breaks/changes then it's likely gonna break for every call to `hash` because of how hashes generally work. so we'd notice in for instance `callable_hash`.